apps/shopify/shopify.star

```diff
@@ -0,0 +1,554 @@
+    else:
+        return None
+
+def get_total_orders_count(store_name, api_token, start_time, since_id):
+    parameters = {"limit": "250", "created_at_min": "{}Z".format(start_time), "status": "any"}
+    if since_id:
+        parameters["since_id"] = since_id
+    else:
+        parameters["since_id"] = "0"
+    print("🕸 Fetching total order count {}".format(parameters))
+    response = make_shopify_request(store_name = store_name, api_token = api_token, endpoint = "orders/count", parameters = parameters)
+    if is_response_error(response):
+        return response
+    else:
+        count = int(response["count"])
+        print("Order Count {}".format(count))
+        return count
+
+def get_orders(store_name, api_token, start_time, since_id):
+    orders = get_cached_orders(start_time, store_name, api_token)
+    if orders:
+        return orders
+    print("Getting orders since {}".format(start_time))
+    orders = []
+    order_count = get_total_orders_count(store_name, api_token, start_time, since_id)
+    if is_response_error(order_count):
+        return order_count
+
+    total_pages = int(order_count / 250) + 1
+    since_id = None
+    for current_page in range(total_pages):
+        chunk = get_chunk_of_orders(store_name, api_token, start_time, since_id)
+        if is_response_error(chunk):
+            return chunk
+        else:
+            orders.extend(chunk)
+            since_id = str(int(chunk[-1]["id"]))
+            print("Setting since_id {}".format(since_id))
+            if len(chunk) == 0:
+                break
+
+    orders = [o for o in orders if o["financial_status"] not in ["refunded", "voided"]]
+    print("Fetched {} orders".format(len(orders)))
+    if orders:
+        cache_orders(orders, start_time, store_name, api_token)
+    return orders
+
+def get_chunk_of_orders(store_name, api_token, start_time, since_id):
+    parameters = {"limit": "250", "created_at_min": "{}Z".format(start_time), "status": "any"}
+    if since_id:
+        parameters["since_id"] = since_id
+        print("Using since_id {}".format(since_id))
+    else:
+        parameters["since_id"] = "0"
+    print("🕸 Fetching orders... {}".format(parameters))
+    response = make_shopify_request(store_name = store_name, api_token = api_token, endpoint = "orders", parameters = parameters)
+    if is_response_error(response):
+        return response
+    else:
+        orders = response["orders"]
+        return orders
+
+def flatten(xss):
+    return [x for xs in xss for x in xs]
+
+def make_shopify_request(store_name, api_token, endpoint, parameters):
+    url = "https://{}.myshopify.com/admin/api/2022-04/{}.json".format(store_name, endpoint)
+    headers = {"Content-Type": "application/json", "X-Shopify-Access-Token": api_token}
+    response = http.get(url = url, params = parameters, headers = headers)
+    if response.status_code == 404:
+        return ERROR_404
+    elif response.status_code == 429:
+        return ERROR_429
```

Comment:
A store with large enough order volumes (and high enough cycle speed) will hit the rate-limit frequently as orders accumulate over the day. Would it be good to not only report the rate-limit, but also respect it by backing off?

Since the rate limit is per store/token, and probably not all that many stores reach that kind of a daily order volume, I suppose it's not critical? WDYT?

Reply:
@matslina is there a way to sleep or wait for an amount of time? I couldn't find that in the docs anywhere.

Reply:
One approach is to write a cache record to mark backoff (e.g. `cache.set("rate-limited", "%d" % backoff, ttl_seconds=backoff)`), and check that before performing any API calls (e.g. `if cache.get("rate-limited"): handle_rate_limit()`).

Reply:
So there's no way to pause an execution by e.g. calling a `sleep()` function, and there's no way to schedule executions. When rate-limit is hit, the current execution has to take some reasonable action (perhaps displaying a "rate-limit" error message.) The app will be executed again the future, as the Tidbyt device cycles through installed apps, offering an opportunity to get fresh data from the API.

The `rate-limited` cache record serves to stop these subsequent executions from hitting the API needlessly for a period of time, allowing the rate limit to be reset. Without it, the app could get into a state where it's calling the API, getting a 429, and then just trying again on the next execution. This can stop a rate-limit from being reset.

Setting an appropriate TTL on the `rate-limited` cache record makes it so that the cache eventually evicts the record. When that happens, new executions will attempt calling the API again, and hopefully stop receiving 429s.

But yeah I'm not sure if this is going to be needed. It'd have to be a pretty large number of daily orders for the rate limit to ever get triggered by Shopify's API.

Reply:
Ya without a true sleep method to respect the wait time sent back by Shopify, I think I would have to really over-engineer something here to support the edge case, at least given what I know now.

Reply:
@matslina if you are good with no action here, then I think its ready.

Reply:
Sure thing. Let's merge.

Just in case you're curious, there's an example of the rate-limit backoff over here: https://github.com/tidbyt/community/blob/main/apps/formula1/formula_1.star#L85